design/20221205-memory-management.md

```diff
@@ -0,0 +1,454 @@
+This mechanism is used by other projects that use client-go controllers, for example [istio](https://github.com/istio/istio/blob/1.16.0/pilot/pkg/status/distribution/state.go#L100-L103).
+The same filtering mechanism is [also available for cert-manager.io resources](https://github.com/cert-manager/cert-manager/blob/v1.10.1/pkg/client/informers/externalversions/factory.go#L63-L69). We shouldn't need to filter what cert-manager.io resoruces we watch though.
+This mechanism seems the most straightforward to use, but currently we don't have a way to identify all resources (secrets) we need to watch using a label or field selector, see [###Secrets].
+
+##### Partial object metadata
+
+Caching only metadata for a given object. This mechanism relies on making list and watch calls against kube apiserver with a `PartialObjectMetadata` header. The apiserver then returns [PartialObjectMetadata](https://github.com/kubernetes/apimachinery/blob/v0.26.0/pkg/apis/meta/v1/types.go#L1425-L1447) instead of an object of a concrete type such as a `Secret`. The `PartialObjectMetadata` only contains the metadata and type information of the object.
+To use this mechanism to ensure that metadata only is being cached for a particular resource type that triggers a reconcile, `ListerWatcher` of the informer for that type needs to use a client that knows how to make calls with `PartialObjectMetadata` header. Also if the reconcile loop can only retrieve `PartialObjectMetadata` types from cache.
+client-go has a [metadata only client](https://github.com/kubernetes/client-go/blob/v0.25.5/metadata/metadata.go#L85-L99) that can be used to get, list and watch with `PartialObjectMetadata`. client-go also has a [metadata informer](https://github.com/kubernetes/client-go/blob/v0.25.5/metadata/metadatainformer/informer.go#L118-L142) that uses the metadata only client to list and watch resources. This informer implements the same [SharedIndexInformer interface](https://github.com/kubernetes/client-go/blob/v0.26.0/tools/cache/shared_informer.go#L219) as the core and cert-manager.io informers that we use currently, so it would fit our existing controller setup. 
+The downside to having metadata only in cache is that if the reconcile loop needs the whole object, it needs to make another call to the kube apiserver to get the actual object. We have a number of reconcile loops that retrieve and parse secret data numerous times, for example [readiness controller](https://github.com/cert-manager/cert-manager/blob/v1.10.1/pkg/controller/certificates/readiness/readiness_controller.go) retrieves and parses `spec.SecretName` secret for a `Certificate` on any event associated with the `Certificate`, any of its `CertificateRequest`s or the `spec.secretName` secret.
+TODO: add which projects have adopted metadata-only watches, especially with client-go informers
```

Comment:
What are the consequences of "retrieving and parsing secret data numerous times"? 

With [API Priority and Fairness](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/)  enabled by default in modern Kubernetes, there is no longer a risk of cert-manager overwhelming the K8S API server.

In the meeting today you explained that there is a risk that making too many requests will result in client side throttling in the K8S client, which may cause slow issuance of certificates, but what if we increased the [client side throttling defaults](https://github.com/cert-manager/cert-manager/pull/3382) or even [Disable client-side rate-limiting when AP&F is enabled](https://github.com/kubernetes/kubernetes/issues/111880)?

In any case, I do think it would be writing a paragraph about why you think it is a problem to "retrieve and parse secret data numerous times" and if possible add some evidence to show what happens when cert-manager makes too many requests per second.
Random thought: Would it be possible to compile a version of cert-manager without any caching, where the listers in the various controllers make API requests rather than cache lookups? And see how it behaves

Refs:
* https://github.com/cert-manager/cert-manager/pull/3382
* https://github.com/kubernetes/kubernetes/issues/111880

Reply:
Regarding "which projects have adopted metadata-only watches" I found these examples:

* https://github.com/kyverno/kyverno/issues/1832
* https://github.com/gardener/cert-management/issues/55
* https://github.com/kubernetes-sigs/cluster-api/issues/3887

Reply:
> What are the consequences of "retrieving and parsing secret data numerous times"?

Additional calls to kube apiserver because of the large number of additional requests. I think the number of calls in this case is too large and there also isn't much of a gain from not caching certificate secrets that usually aren't that big anyway.
In my opinion it does not matter if the additional calls result in client-side throttling or AP&F throttling or even if users can somehow configure it in such a way that cert-manager is 'allowed' to make more requests.
These calls aren't free for users in any case - either it's a slowdown of issuance or less availability for other cluster tenants and there isn't much of a gain memory-wise because the certificate secrets aren't that big so caching them isn't that expensive so I think we should not stop caching secrets that we know we need.

> Random thought: Would it be possible to compile a version of cert-manager without any caching, where the listers in the various controllers make API requests rather than cache lookups? And see how it behaves

I've modified the prototype for this design to do that https://github.com/irbekrm/cert-manager/tree/just_partial and added a section that describes this approach to 'Alternatives' https://github.com/irbekrm/cert-manager/blob/memory_design/design/20221205-memory-management.md#use-partialmetadata-only